clippy_utils/src/diagnostics.rs

```diff
@@ -152,13 +152,51 @@ where
     });
 }
 
+/// Like [`span_lint`], but allows providing the `HirId` of the node that caused us to emit this
+/// lint.
+///
+/// The `HirId` is used for checking lint level attributes.
+///
+/// For example:
+/// ```ignore
+/// fn f() { /* '1 */
+///
+///     #[allow(clippy::some_lint)]
+///     let _x = /* '2 */;
+/// }
+/// ```
+/// If `some_lint` does its analysis in `LintPass::check_fn` (at `'1`) and emits a lint at `'2`
+/// using [`span_lint`], then allowing the lint at `'2` as attempted in the snippet will not work!
+/// Even though that is where the warning points at, which would be confusing to users.
```

Comment:
The naming `'1` can be a bit confusing IMO, since that's how lifetimes are usually called. Maybe `<node_1>` or `<expr_1>` would be better?

Reply:
Agreed, that does sound better👍